### PR TITLE
Allow to disable events publisher

### PR DESCRIPTION
**Pull request by VKAlwaysWin**
**Describe the change**
Allow to disable LifeCycleEventsPublisher

**Breaking change**
No breaking changes.

### CODE REVIEW

**Review by danielgerlag — COMMENTED** [review]
What problem is this trying to solve?

**Comment by VKAlwaysWin** [comment]
@danielgerlag Performance mainly. Those events used only in case if you have OnLifeCycleEvent and in all other cases it is not used. Please correct me if I wrong here.

In case of really big load 10+ workflows per second with a lot of steps GetConsumingEnumerable for BlockingCollection is taking couple of percents (~ 3.5% in my case) of execution time. 

And this PR allowing to save couple of ms per workflow if you don't need to track LifeCycle events. Plus it shouldn't affect current default settings and implementation.

Does that make sense ?

File: src/WorkflowCore/Services/ErrorHandlers/SuspendHandler.cs [hunk 1]
```diff
@@ -21,7 +21,7 @@ public SuspendHandler(ILifeCycleEventPublisher eventPublisher, IDateTimeProvider
         public void Handle(WorkflowInstance workflow, WorkflowDefinition def, ExecutionPointer pointer, WorkflowStep step, Exception exception, Queue<ExecutionPointer> bubbleUpQueue)
         {
             workflow.Status = WorkflowStatus.Suspended;
-            _eventPublisher.PublishNotification(new WorkflowSuspended
+            _eventPublisher?.PublishNotification(new WorkflowSuspended
```
Comment (danielgerlag) [inline src/WorkflowCore/Services/ErrorHandlers/SuspendHandler.cs:24]: The fact that the publisher can be disabled is leaking into where it is consumed, this creates inappropriate coupling.

File: src/WorkflowCore/Services/ErrorHandlers/TerminateHandler.cs [hunk 2]
```diff
@@ -23,7 +23,7 @@ public void Handle(WorkflowInstance workflow, WorkflowDefinition def, ExecutionP
             workflow.Status = WorkflowStatus.Terminated;
             workflow.CompleteTime = _dateTimeProvider.UtcNow;
 
-            _eventPublisher.PublishNotification(new WorkflowTerminated
+            _eventPublisher?.PublishNotification(new WorkflowTerminated
```
Comment (danielgerlag) [inline src/WorkflowCore/Services/ErrorHandlers/TerminateHandler.cs:26]: same here

File: src/WorkflowCore/Services/ExecutionResultProcessor.cs [hunk 3]
```diff
@@ -77,7 +77,7 @@ public void ProcessExecutionResult(WorkflowInstance workflow, WorkflowDefinition
                     subsequent.Active = true;
                 }
 
-                _eventPublisher.PublishNotification(new StepCompleted
+                _eventPublisher?.PublishNotification(new StepCompleted
```
Comment (danielgerlag) [inline src/WorkflowCore/Services/ExecutionResultProcessor.cs:80]: same here

File: src/WorkflowCore/Services/ExecutionResultProcessor.cs [hunk 4]
```diff
@@ -102,7 +102,7 @@ public void ProcessExecutionResult(WorkflowInstance workflow, WorkflowDefinition
 
         public void HandleStepException(WorkflowInstance workflow, WorkflowDefinition def, ExecutionPointer pointer, WorkflowStep step, Exception exception)
         {
-            _eventPublisher.PublishNotification(new WorkflowError
+            _eventPublisher?.PublishNotification(new WorkflowError
```
Comment (danielgerlag) [inline src/WorkflowCore/Services/ExecutionResultProcessor.cs:105]: same here

File: src/WorkflowCore/Services/WorkflowExecutor.cs [hunk 5]
```diff
@@ -123,7 +123,7 @@ private bool InitializeStep(WorkflowInstance workflow, WorkflowStep step, Workfl
             if (pointer.Status != PointerStatus.Running)
             {
                 pointer.Status = PointerStatus.Running;
-                _publisher.PublishNotification(new StepStarted
+                _publisher?.PublishNotification(new StepStarted
```
Comment (danielgerlag) [inline src/WorkflowCore/Services/WorkflowExecutor.cs:126]: same here

File: src/WorkflowCore/Services/WorkflowExecutor.cs [hunk 6]
```diff
@@ -264,7 +264,7 @@ private async Task DetermineNextExecutionTime(WorkflowInstance workflow, Workflo
                 await middlewareRunner.RunPostMiddleware(workflow, def);
             }
 
-            _publisher.PublishNotification(new WorkflowCompleted
+            _publisher?.PublishNotification(new WorkflowCompleted
```
Comment (danielgerlag) [inline src/WorkflowCore/Services/WorkflowExecutor.cs:267]: same here

File: src/WorkflowCore/Services/ErrorHandlers/SuspendHandler.cs (hunk 1 above)
Comment (danielgerlag) [reply src/WorkflowCore/Services/ErrorHandlers/SuspendHandler.cs:24]: Why not just put the check in the publisher?

**Comment by VKAlwaysWin** [comment]
@danielgerlag added check before adding into outbox. Will that work for you ?
Also added unit test to cover this use case and improved existing ones.

**Comment by VKAlwaysWin** [comment]
@danielgerlag any chance to get this merged ?

**Pull request merged**